geo/src/algorithm/line_measures/metric_spaces/euclidean/distance.rs

```diff
@@ -0,0 +1,1087 @@
+                      Geometry::Triangle(triangle) => Self::distance(origin, triangle),
+                  }
+              }
+          }
+          symmetric_distance_impl!(GeoFloat, &Geometry<F>, $target);
+      )*
+  };
+}
+
+impl_euclidean_distance_for_geometry_and_variant!([&Point<F>, &MultiPoint<F>, &Line<F>, &LineString<F>, &MultiLineString<F>, &Polygon<F>, &MultiPolygon<F>, &Triangle<F>, &Rect<F>, &GeometryCollection<F>]);
+
+impl<F: GeoFloat> Distance<F, &Geometry<F>, &Geometry<F>> for Euclidean {
+    fn distance(origin: &Geometry<F>, destination: &Geometry<F>) -> F {
+        match origin {
+            Geometry::Point(point) => Self::distance(point, destination),
+            Geometry::Line(line) => Self::distance(line, destination),
+            Geometry::LineString(line_string) => Self::distance(line_string, destination),
+            Geometry::Polygon(polygon) => Self::distance(polygon, destination),
+            Geometry::MultiPoint(multi_point) => Self::distance(multi_point, destination),
+            Geometry::MultiLineString(multi_line_string) => {
+                Self::distance(multi_line_string, destination)
+            }
+            Geometry::MultiPolygon(multi_polygon) => Self::distance(multi_polygon, destination),
+            Geometry::GeometryCollection(geometry_collection) => {
+                Self::distance(geometry_collection, destination)
+            }
+            Geometry::Rect(rect) => Self::distance(rect, destination),
+            Geometry::Triangle(triangle) => Self::distance(triangle, destination),
+        }
+    }
+}
+
+// ┌───────────────────────────┐
+// │ Implementations utilities │
+// └───────────────────────────┘
+
+/// Uses an R* tree and nearest-neighbour lookups to calculate minimum distances
+// This is somewhat slow and memory-inefficient, but certainly better than quadratic time
+fn nearest_neighbour_distance<F: GeoFloat>(geom1: &LineString<F>, geom2: &LineString<F>) -> F {
+    let tree_a = RTree::bulk_load(geom1.lines().map(CachedEnvelope::new).collect());
+    let tree_b = RTree::bulk_load(geom2.lines().map(CachedEnvelope::new).collect());
+    // Return minimum distance between all geom a points and geom b lines, and all geom b points and geom a lines
+    geom2
+        .points()
+        .fold(Bounded::max_value(), |acc: F, point| {
+            let nearest = tree_a.nearest_neighbor(&point).unwrap();
+            acc.min(Euclidean::distance(nearest as &Line<F>, &point))
+        })
+        .min(geom1.points().fold(Bounded::max_value(), |acc, point| {
+            let nearest = tree_b.nearest_neighbor(&point).unwrap();
+            acc.min(Euclidean::distance(nearest as &Line<F>, &point))
+        }))
+}
+
+fn ring_contains_coord<T: GeoNum>(ring: &LineString<T>, c: Coord<T>) -> bool {
+    match coord_pos_relative_to_ring(c, ring) {
+        CoordPos::Inside => true,
+        CoordPos::OnBoundary | CoordPos::Outside => false,
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
These are the tests copied from the legacy euclidean_distance trait, updated to use the new trait. I didn't review them very carefully, except to make sure that they still passed.